dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/FindSpinLoops.java

```diff
@@ -63,23 +63,31 @@ public void run(Program program) {
         logger.info("# of spinloops: {}", spinloops);
 }
 
+	// This method assumes the following implementation of await_while
+	// #define await_while(cond)                                                  \
+    // for (int tmp = (__VERIFIER_loop_begin(), 0); __VERIFIER_spin_start(),  \
+    //     tmp = cond, __VERIFIER_spin_end(!tmp), tmp;)
     private void markAnnotatedSpinLoops(Thread t){
-    	Event pred = t.getEntry();
-        Event curr = pred.getSuccessor();
+        Event curr = t.getEntry();
         while (curr != null) {
-        	if(curr instanceof LoopEnd) {
-        		// This assume the following implementation of await_while
-        		// #define await_while(cond)                                                  \
-        	    // for (int tmp = (__VERIFIER_loop_begin(), 0); __VERIFIER_spin_start(),  \
-        	    //     tmp = cond, __VERIFIER_spin_end(!tmp), tmp;)
-        		Event spinloop = curr.getSuccessors().stream().filter(e -> e instanceof CondJump && ((CondJump)e).isGoto()).findFirst().get();
-        		spinloop.addFilters(Tag.SPINLOOP, Tag.NOOPT);
-                ((CondJump)spinloop).getLabel().addFilters(Tag.SPINLOOP, Tag.NOOPT);
-        		spinloops++;
-        	}
+            // Find start of spinloop.
+            if (curr.getSuccessor() instanceof LoopStart) {
+                Label label = (Label) curr;
```

Comment:
This only works if the IR has the form
```
loopLabel:
#LoopStart
```
This might be the case when using the `await_while` construct but I think it is a bit risky. If the user explicitly places a `__VERIFIER_spin_start` not directly after the label, the code will trigger a `ClassCastException`.
I think we should at least check if `curr` is a `Label`, and if it is not, ignore the `LoopStart` and log a warning that the `LoopStart` event is misplaced.

Also, wouldn't it be nicer to have a single `__VERIFIER_spin_loop` that is placed somewhere before the start of the spinloop (like `__VERIFIER_loop_bound`)? Cause right now, I think it is impossible to specify a spinloop without using `await_while`, which might also be a good thing...

Reply:
I agree about the possibly unsafe cast. I'll fix that.

> Also, wouldn't it be nicer to have a single `__VERIFIER_spin_loop` that is placed somewhere before the start of the spinloop (like `__VERIFIER_loop_bound`)? Cause right now, I think it is impossible to specify a spinloop without using `await_while`, which might also be a good thing...

The purpose of adding the three `__VERIFIER_X` methods related to spinloops was to support this annotation in a way which is consistent with GenMC and thus all benchmarks from the DRC would work out of the box. Since right now we don't have other use cases (where the general idea you propose would work better) and changing this would not work with e.g. the CNA code, I think we should keep using this one for the moment.